In [133]:
import pandas as pd
import numpy as np
import matplotlib as plt
import seaborn as sns

In [134]:
%matplotlib inline

In [135]:
#variable to use for import data path
outpatient_fp='../data/MUP_OHP_R19_P04_V10_D17_APC_Provider.xlsx'

In [136]:
#import data with selected columns, cut off 1st four rows and use row 5 as culumn names, 
# convert two numeric columns to strings to preserve any leading zeros
outpatient=pd.read_excel(
    outpatient_fp,
    header=5,
    usecols=['Provider ID','Provider Name','Provider City','Provider\nState','Provider\nZip Code',
            'APC','APC\nDescription','Beneficiaries','Comprehensive APC\nServices',
             'Average\nMedicare\nAllowed\nAmount','Average\nMedicare\nPayment\nAmount'],
    dtype={'Provider\nZip Code':'str','APC':'str'}
    )

In [137]:
#lower case column lables
outpatient.columns=map(str.lower,outpatient.columns)

In [138]:
#replace blank spaces with underscores(_)
outpatient.columns=outpatient.columns.str.strip().str.replace(' ','_')

In [139]:
#replace blank new line character(\n) with underscores(_)
outpatient.columns=outpatient.columns.str.strip().str.replace('\n','_')                                                      

In [140]:
outpatient

provider_id                             provider_name  provider_city  \
0            10001          Southeast Alabama Medical Center         Dothan   
1            10001          Southeast Alabama Medical Center         Dothan   
2            10001          Southeast Alabama Medical Center         Dothan   
3            10001          Southeast Alabama Medical Center         Dothan   
4            10001          Southeast Alabama Medical Center         Dothan   
...            ...                                       ...            ...   
61774       670122  Houston Methodist The Woodlands Hospital  The Woodlands   
61775       670122  Houston Methodist The Woodlands Hospital  The Woodlands   
61776       670122  Houston Methodist The Woodlands Hospital  The Woodlands   
61777       670122  Houston Methodist The Woodlands Hospital  The Woodlands   
61778       670122  Houston Methodist The Woodlands Hospital  The Woodlands   

      provider_state provider_zip_code   apc  \
0                 AL             36301  5072   
1                 AL             36301  5073   
2                 AL             36301  5091   
3                 AL             36301  5092   
4                 AL             36301  5112   
...              ...               ...   ...   
61774             TX             77385  5361   
61775             TX             77385  5373   
61776             TX             77385  5375   
61777             TX             77385  5431   
61778             TX             77385  8011   

                                         apc_description  beneficiaries  \
0        Level 2 Excision/ Biopsy/ Incision and Drainage          249.0   
1        Level 3 Excision/ Biopsy/ Incision and Drainage           52.0   
2      Level 1 Breast/Lymphatic Surgery and Related P...           26.0   
3      Level 2 Breast/Lymphatic Surgery and Related P...           23.0   
4                     Level 2 Musculoskeletal Procedures           17.0   
...                                                  ...            ...   
61774           Level 1 Laparoscopy and Related Services           18.0   
61775               Level 3 Urology and Related Services           11.0   
61776               Level 5 Urology and Related Services           20.0   
61777                           Level 1 Nerve Procedures            NaN   
61778                 Comprehensive Observation Services          156.0   

       comprehensive_apc_services  average_medicare_allowed_amount  \
0                             259                      1038.454672   
1                              53                      1792.596792   
2                              27                      2113.580000   
3                              23                      3737.140000   
4                              17                      1029.460000   
...                           ...                              ...   
61774                          18                      3866.822778   
61775                          11                      1469.871818   
61776                          21                      3356.890000   
61777                          12                      1506.380000   
61778                         165                      2141.540000   

       average_medicare_payment_amount  
0                           826.277954  
1                          1423.246792  
2                          1683.990000  
3                          2977.550000  
4                           820.210000  
...                                ...  
61774                      3044.482778  
61775                      1147.781818  
61776                      2674.590000  
61777                      1200.200000  
61778                      1706.270000  

[61779 rows x 11 columns]

In [141]:
outpatient.isnull().sum()

provider_id                          0
provider_name                        0
provider_city                        0
provider_state                       0
provider_zip_code                    0
apc                                  0
apc_description                      0
beneficiaries                      997
comprehensive_apc_services           0
average_medicare_allowed_amount      0
average_medicare_payment_amount      0
dtype: int64

In [143]:
#rename column 'provider_zip_code' to 'zip_code' to elimate doubling of zip code column n merged data frame
outpatient=outpatient.rename(columns={'provider_zip_code':'zip_code'}).dropna()
outpatient

provider_id                             provider_name  provider_city  \
0            10001          Southeast Alabama Medical Center         Dothan   
1            10001          Southeast Alabama Medical Center         Dothan   
2            10001          Southeast Alabama Medical Center         Dothan   
3            10001          Southeast Alabama Medical Center         Dothan   
4            10001          Southeast Alabama Medical Center         Dothan   
...            ...                                       ...            ...   
61773       670122  Houston Methodist The Woodlands Hospital  The Woodlands   
61774       670122  Houston Methodist The Woodlands Hospital  The Woodlands   
61775       670122  Houston Methodist The Woodlands Hospital  The Woodlands   
61776       670122  Houston Methodist The Woodlands Hospital  The Woodlands   
61778       670122  Houston Methodist The Woodlands Hospital  The Woodlands   

      provider_state zip_code   apc  \
0                 AL    36301  5072   
1                 AL    36301  5073   
2                 AL    36301  5091   
3                 AL    36301  5092   
4                 AL    36301  5112   
...              ...      ...   ...   
61773             TX    77385  5302   
61774             TX    77385  5361   
61775             TX    77385  5373   
61776             TX    77385  5375   
61778             TX    77385  8011   

                                         apc_description  beneficiaries  \
0        Level 2 Excision/ Biopsy/ Incision and Drainage          249.0   
1        Level 3 Excision/ Biopsy/ Incision and Drainage           52.0   
2      Level 1 Breast/Lymphatic Surgery and Related P...           26.0   
3      Level 2 Breast/Lymphatic Surgery and Related P...           23.0   
4                     Level 2 Musculoskeletal Procedures           17.0   
...                                                  ...            ...   
61773                        Level 2 Upper GI Procedures           17.0   
61774           Level 1 Laparoscopy and Related Services           18.0   
61775               Level 3 Urology and Related Services           11.0   
61776               Level 5 Urology and Related Services           20.0   
61778                 Comprehensive Observation Services          156.0   

       comprehensive_apc_services  average_medicare_allowed_amount  \
0                             259                      1038.454672   
1                              53                      1792.596792   
2                              27                      2113.580000   
3                              23                      3737.140000   
4                              17                      1029.460000   
...                           ...                              ...   
61773                          18                      1286.130000   
61774                          18                      3866.822778   
61775                          11                      1469.871818   
61776                          21                      3356.890000   
61778                         165                      2141.540000   

       average_medicare_payment_amount  
0                           826.277954  
1                          1423.246792  
2                          1683.990000  
3                          2977.550000  
4                           820.210000  
...                                ...  
61773                      1024.720000  
61774                      3044.482778  
61775                      1147.781818  
61776                      2674.590000  
61778                      1706.270000  

[60782 rows x 11 columns]

In [144]:
outpatient.columns

Index(['provider_id', 'provider_name', 'provider_city', 'provider_state',
       'zip_code', 'apc', 'apc_description', 'beneficiaries',
       'comprehensive_apc_services', 'average_medicare_allowed_amount',
       'average_medicare_payment_amount'],
      dtype='object')

In [145]:
outpatient.shape

(60782, 11)

In [146]:
outpatient.dtypes

provider_id                          int64
provider_name                       object
provider_city                       object
provider_state                      object
zip_code                            object
apc                                 object
apc_description                     object
beneficiaries                      float64
comprehensive_apc_services           int64
average_medicare_allowed_amount    float64
average_medicare_payment_amount    float64
dtype: object

In [147]:
outpatient.isnull().sum()

provider_id                        0
provider_name                      0
provider_city                      0
provider_state                     0
zip_code                           0
apc                                0
apc_description                    0
beneficiaries                      0
comprehensive_apc_services         0
average_medicare_allowed_amount    0
average_medicare_payment_amount    0
dtype: int64

In [82]:
#import zip_cbsa.cvs
cbsa=pd.read_csv(
    '../data/zip_cbsa.csv',
    #convert numeric data to strings to preserve leading zeros
    dtype={'zip_code':'str','cbsa':'str'})

In [148]:
cbsa

zip_code   cbsa
0        00501  35620
1        00601  38660
2        00602  10380
3        00603  10380
4        00604  10380
...        ...    ...
47410    99925  99999
47411    99926  99999
47412    99927  99999
47413    99928  28540
47414    99929  99999

[47415 rows x 2 columns]

In [149]:
#cbsa=cbsa.rename(columns={'zip_code':'r_zip', 'cbsa':'r_cbsa'})
#cbsa

In [150]:
cbsa.dtypes

zip_code    object
cbsa        object
dtype: object

In [200]:
cbsa_nash=cbsa[cbsa['zip_code']=='37205']
cbsa_nash

zip_code   cbsa
17171    37205  34980

In [152]:
#Create dictionary from cbsa
 #df.set_index('ID').T.to_dict('list')
#cbsa_dict=cbsa.set_index('zip_code').T.to_dict('list')

In [153]:
#print(cbsa_dict)

In [154]:
#use cbsa_dict to look up CBSA's for Zip Codes
#df['new column name'] = df.colunm.map(dictionary)
#outpatient['cbsa']=outpatient.zip_code.map(cbsa_dict)

In [155]:
#outpatient

In [156]:
#outpatient.shape

In [157]:
#merged_table = merged_table.drop_duplicates(subset='column')

In [158]:
# pd.merge(df,df1,how='outer').groupby([0,1],as_index=False)[2].sum()

In [106]:
#join zip_cbsa
outpatient_cbsa_zip=pd.merge(outpatient,cbsa, left_on='zip_code',right_on="zip_code",how='left')
#outpatient_cbsa_zip=pd.merge(outpatient, cbsa, on='zip_code').groupby(['apc','beneficiaries'],as_index=False)['zip_code'].sum()#.drop_duplicates(subset= "zip_code")

In [159]:
outpatient_cbsa_zip

provider_id                             provider_name  provider_city  \
0            10001          Southeast Alabama Medical Center         Dothan   
1            10001          Southeast Alabama Medical Center         Dothan   
2            10001          Southeast Alabama Medical Center         Dothan   
3            10001          Southeast Alabama Medical Center         Dothan   
4            10001          Southeast Alabama Medical Center         Dothan   
...            ...                                       ...            ...   
69607       670122  Houston Methodist The Woodlands Hospital  The Woodlands   
69608       670122  Houston Methodist The Woodlands Hospital  The Woodlands   
69609       670122  Houston Methodist The Woodlands Hospital  The Woodlands   
69610       670122  Houston Methodist The Woodlands Hospital  The Woodlands   
69611       670122  Houston Methodist The Woodlands Hospital  The Woodlands   

      provider_state zip_code   apc  \
0                 AL    36301  5072   
1                 AL    36301  5073   
2                 AL    36301  5091   
3                 AL    36301  5092   
4                 AL    36301  5112   
...              ...      ...   ...   
69607             TX    77385  5302   
69608             TX    77385  5361   
69609             TX    77385  5373   
69610             TX    77385  5375   
69611             TX    77385  8011   

                                         apc_description  beneficiaries  \
0        Level 2 Excision/ Biopsy/ Incision and Drainage          249.0   
1        Level 3 Excision/ Biopsy/ Incision and Drainage           52.0   
2      Level 1 Breast/Lymphatic Surgery and Related P...           26.0   
3      Level 2 Breast/Lymphatic Surgery and Related P...           23.0   
4                     Level 2 Musculoskeletal Procedures           17.0   
...                                                  ...            ...   
69607                        Level 2 Upper GI Procedures           17.0   
69608           Level 1 Laparoscopy and Related Services           18.0   
69609               Level 3 Urology and Related Services           11.0   
69610               Level 5 Urology and Related Services           20.0   
69611                 Comprehensive Observation Services          156.0   

       comprehensive_apc_services  average_medicare_allowed_amount  \
0                             259                      1038.454672   
1                              53                      1792.596792   
2                              27                      2113.580000   
3                              23                      3737.140000   
4                              17                      1029.460000   
...                           ...                              ...   
69607                          18                      1286.130000   
69608                          18                      3866.822778   
69609                          11                      1469.871818   
69610                          21                      3356.890000   
69611                         165                      2141.540000   

       average_medicare_payment_amount   cbsa  
0                           826.277954  20020  
1                          1423.246792  20020  
2                          1683.990000  20020  
3                          2977.550000  20020  
4                           820.210000  20020  
...                                ...    ...  
69607                      1024.720000  26420  
69608                      3044.482778  26420  
69609                      1147.781818  26420  
69610                      2674.590000  26420  
69611                      1706.270000  26420  

[69612 rows x 12 columns]

In [160]:
cbsa.shape, outpatient.shape

((47415, 2), (60782, 11))

In [161]:
#df.append(df2, ignore_index=True)
#outpatient_cbsa_zip=outpatient.append(cbsa, ignore_index=True)

In [162]:
#add refernce zip code (r_zip) and reference cbsa (ref_cbsa) columns to outpatient, rename 'outpatient_cbsa_zip'
#df_new = pd.concat([df1, df2], axis=1)
#outpatient_cbsa_zip=pd.concat([outpatient,cbsa],axis=1)

outpatient has 60782 rows, joining with zip_cbsa added 8830 rows (69612 total) . Zip codes and CBSA's do not neccesarily share boundries. While CBSA's contain multiple zip codes, a particular zip code may also cross two CBSA's

In [163]:
outpatient_cbsa_zip.isnull().sum()

provider_id                           0
provider_name                         0
provider_city                         0
provider_state                        0
zip_code                              0
apc                                   0
apc_description                       0
beneficiaries                         0
comprehensive_apc_services            0
average_medicare_allowed_amount       0
average_medicare_payment_amount       0
cbsa                               4877
dtype: int64

In [167]:
outpatient_cbsa_zip.shape

(69612, 12)

In [185]:
#import hospitals to merge to outpatient to add hcpcs column, creating data frame apc_pcs
apc_hcpcs=pd.read_csv('../data/hospital.csv',
                      #convert numeric data to strings to preserve leading zeros
                      dtype={'hcpcs_code':'str','apc':'str'})
apc_hcpcs

hcpcs_code              short_descriptor   apc  \
0         10005       Fna bx w/us gdn 1st les  5071   
1         10007    Fna bx w/fluor gdn 1st les  5071   
2         10009       Fna bx w/ct gdn 1st les  5071   
3         10011       Fna bx w/mr gdn 1st les  5071   
4         10021    Fna bx w/o img gdn 1st les  5052   
...         ...                           ...   ...   
5929      Q9950  Inj sulf hexa lipid microsph  9085   
5930      Q9968         Visualization adjunct  1446   
5931      Q9969    Non-heu tc-99m add-on/dose  1442   
5932      Q9991  Buprenorph xr 100 mg or less  9073   
5933      Q9992  Buprenorphine xr over 100 mg  9239   

     national_unadjusted_copayment  minimum_unadjusted_copayment  
0                                .                      122.0100  
1                                .                      122.0100  
2                                .                      122.0100  
3                                .                      122.0100  
4                                .                       63.9100  
...                            ...                           ...  
5929                             0                        0.0000  
5930                             .                        1.2224  
5931                             .                        2.0000  
5932                             .                      334.4908  
5933                             .                      334.4908  

[5934 rows x 5 columns]

In [186]:
apc_hcpcs.dtypes

hcpcs_code                        object
short_descriptor                  object
apc                               object
national_unadjusted_copayment     object
minimum_unadjusted_copayment     float64
dtype: object

In [198]:
apc_5302_hcps=apc_hcpcs[apc_hcpcs['apc']=='5302']
apc_5302_hcps.hcpcs_code

1875    32556
2339    43191
2340    43192
2341    43193
2342    43194
2348    43201
2349    43202
2350    43204
2351    43205
2352    43206
2354    43211
2356    43213
2357    43214
2358    43215
2359    43216
2360    43217
2361    43220
2362    43226
2363    43227
2365    43231
2366    43232
2367    43233
2370    43237
2371    43238
2374    43241
2375    43242
2376    43243
2377    43244
2378    43245
2379    43246
2382    43249
2383    43250
2384    43251
2386    43253
2387    43254
2388    43255
2390    43259
2398    43270
2412    43453
2434    43830
2447    44360
2448    44361
2449    44363
2450    44364
2451    44365
2452    44366
2453    44369
2455    44372
2456    44373
2457    44376
2458    44377
2459    44378
2462    44381
2664    49423
2667    49436
2668    49440
2669    49441
2671    49446
5170    0355T
Name: hcpcs_code, dtype: object

In [197]:
apc_5302_hcps.shape

(59, 5)

In [191]:
apc_5302_hcps.isnull().sum()

hcpcs_code                       0
short_descriptor                 0
apc                              0
national_unadjusted_copayment    0
minimum_unadjusted_copayment     0
dtype: int64

In [177]:
#In new df = df[df["column"] condition]
outpatient_apc_5302=outpatient_cbsa_zip[outpatient_cbsa_zip['apc']=='5302']
outpatient_apc_5302

provider_id                                      provider_name  \
22           10001                   Southeast Alabama Medical Center   
114          10005                      Marshall Medical Center South   
115          10005                      Marshall Medical Center South   
116          10005                      Marshall Medical Center South   
117          10005                      Marshall Medical Center South   
...            ...                                                ...   
69523       670102                              Ad Hospital East, Llc   
69527       670103                              Medical City Alliance   
69549       670108  Baylor Scott & White Medical Center - Marble F...   
69550       670108  Baylor Scott & White Medical Center - Marble F...   
69607       670122           Houston Methodist The Woodlands Hospital   

       provider_city provider_state zip_code   apc  \
22            Dothan             AL    36301  5302   
114             Boaz             AL    35957  5302   
115             Boaz             AL    35957  5302   
116             Boaz             AL    35957  5302   
117             Boaz             AL    35957  5302   
...              ...            ...      ...   ...   
69523        Houston             TX    77015  5302   
69527     Fort Worth             TX    76177  5302   
69549   Marble Falls             TX    78654  5302   
69550   Marble Falls             TX    78654  5302   
69607  The Woodlands             TX    77385  5302   

                   apc_description  beneficiaries  comprehensive_apc_services  \
22     Level 2 Upper GI Procedures          165.0                         177   
114    Level 2 Upper GI Procedures           48.0                          64   
115    Level 2 Upper GI Procedures           48.0                          64   
116    Level 2 Upper GI Procedures           48.0                          64   
117    Level 2 Upper GI Procedures           48.0                          64   
...                            ...            ...                         ...   
69523  Level 2 Upper GI Procedures           16.0                          16   
69527  Level 2 Upper GI Procedures           20.0                          25   
69549  Level 2 Upper GI Procedures           27.0                          33   
69550  Level 2 Upper GI Procedures           27.0                          33   
69607  Level 2 Upper GI Procedures           17.0                          18   

       average_medicare_allowed_amount  average_medicare_payment_amount   cbsa  
22                         1123.669040                       894.249040  20020  
114                        1148.718437                       912.358438  22840  
115                        1148.718437                       912.358438  13820  
116                        1148.718437                       912.358438  23460  
117                        1148.718437                       912.358438  10700  
...                                ...                              ...    ...  
69523                      1260.400000                      1004.220000  26420  
69527                      1266.740000                      1009.270000  19100  
69549                      1137.490000                       906.290000  12420  
69550                      1137.490000                       906.290000  99999  
69607                      1286.130000                      1024.720000  26420  

[2592 rows x 12 columns]